ecstask/ecs_task.go

```diff
@@ -397,44 +402,121 @@ func removeAFromB(a, b []string) []string {
 	return diff
 }
 
-func deregisterTaskDefinitions(svc *ecs.ECS, taskDefinitionArns []string, parallel int) {
-	arnsChan := make(chan string, len(taskDefinitionArns))
+// Job carries information through a Job channel.
+type Job struct {
+	Arn string
+}
+
+// Result carries information through a Result channel.
+type Result struct {
+	Arn string
+	Err error
+}
+
+func deregisterTaskDefinitions(svc *ecs.ECS, taskDefinitionArns []string, parallel int, debug bool) {
+	worker := func(wg *sync.WaitGroup, jobsChan <-chan Job, resultsChan chan<- Result) {
+		for job := range jobsChan {
+			input := &ecs.DeregisterTaskDefinitionInput{
+				TaskDefinition: aws.String(job.Arn),
+			}
+
+			_, err := svc.DeregisterTaskDefinition(input)
 
-	deregisterTaskDefinition := func(arn string) {
-		_, err := svc.DeregisterTaskDefinition(&ecs.DeregisterTaskDefinitionInput{
-			TaskDefinition: aws.String(arn),
-		})
-		if err != nil {
-			fmt.Println("Error deregistering task definition:", err)
+			result := Result{Arn: job.Arn, Err: err}
+			resultsChan <- result
 		}
+
+		wg.Done()
 	}
 
-	worker := func(wg *sync.WaitGroup) {
-		for arn := range arnsChan {
-			deregisterTaskDefinition(arn)
+	dispatcher := func(wg *sync.WaitGroup, arns []string, parallel int, jobsChan chan Job, resultsChan chan Result) {
+		jobs := stack.New()
+		for _, arn := range arns {
+			jobs.Push(Job{arn})
+		}
+
+		var completedJobs int
+
+		preload := 1
+		if parallel > 1 {
+			preload = parallel - 1
+		}
+
+		for i := 0; i < preload; i++ {
+			jobsChan <- jobs.Pop().(Job)
+		}
+
+		b := &backoff.Backoff{
+			Min:    100 * time.Millisecond,
+			Max:    2 * time.Minute,
+			Jitter: true,
+		}
+
+		for result := range resultsChan {
+			if result.Err != nil {
+				if !isThrottlingError(result.Err) {
+					fmt.Println("\nError deregistering task definition,", result.Err)
+					close(jobsChan)
+					close(resultsChan)
+					wg.Done()
+					os.Exit(1)
```

Comment:
Little confused by the use of `os.Exit()` here, what's the intention?

Reply:
I want to stop processing jobs if we run into an unexpected error and signal as such, but maybe just a simple `return` would do the job better.